src/escape.rs

```diff
@@ -131,6 +155,116 @@ where
         }
         i += 1;
     }
-    w.write_all(&bytes[mark..])?;
+    w.write_all(&bytes[mark..])
+}
+
+/// Calls the given function with the index of every byte that is in
+/// [", &, <, >] and for no other byte.
+#[cfg(all(target_arch = "x86_64", feature="simd"))]
+fn scan_simd<F>(bytes: &[u8], mut offset: usize, mut callback: F) -> io::Result<()>
+    where F: FnMut(usize) -> io::Result<()>
+{
+    if bytes.is_empty() {
+        // we need this special case for when we are passed an empty subslice
+        // that is exactly on a page boundary. we would try to read on a page
+        // that is potentially not ours and crash
+        return Ok(());
+    }
+
+    let lower_vec = unsafe { _mm_set_epi8( 0, 62, 0, 60, 0, 0, 0, 0, 0, 38, 0, 0, 0, 34, 0, 127 ) };
+    let upperbound = bytes.len().saturating_sub(16);
+
+    while offset < upperbound {
+        let mut mask = unsafe {
+            let raw_ptr = transmute(bytes.as_ptr().offset(offset as isize));
+            let v = _mm_loadu_si128(raw_ptr);
+            let expected = _mm_shuffle_epi8(lower_vec, v);
+            let matches = _mm_cmpeq_epi8(expected, v);
+            _mm_movemask_epi8(matches)
```

Comment:
Hmmm, does this lead to false positives? Since you're checking only the nybbles of `&"<>`, I think it would be possible for non-ASCII bytes to be incorrectly reported as a position here. For example, I think _any_ byte whose 4 least significant binary digits is `0010` will be reported as a `"` by this code.

Reply:
Are you sure? I believe it is sound. Indeed any byte with lower nibble `0010` will get mapped to `"`, but only `"` will test positive in the subsequent comparison?

Reply:
Oh, I see! Gotya. Clever. So I'll turn this into another piece of feedback, but obviously you should defer to @raphlinus here since I'm just a driveby commenter. Namely, I would suggest:

1. Adding some high level docs inside this function describing how this routine works. If I may be allowed to continue to link to memchr... :-) https://github.com/BurntSushi/rust-memchr/blob/d70ea8a1dedd89be028e11f09c99423d4393e26f/src/x86/sse2.rs#L19
2. Add a test that specifically checks that false positives are prevented, e.g., use a byte with the same nybble as `"`. (Maybe you already have this test, but I can't tell easily.)

Reply:
The feedback is certainly appreciated man. Thank you. I will take it to heart :-) :bowing_man: